drivers/sps30/sps30.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (C) 2020 Michel Rottleuthner
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_sps30
+ * @brief       Device driver for the Sensirion SPS30 particulate matter sensor
+ * @author      Michel Rottleuthner <michel.rottleuthner@haw-hamburg.de>
+ * @file
+ */
+#define LOG_LEVEL LOG_DEBUG
+#include "log.h"
+
+#include <errno.h>
+#include <string.h>
+
+#include "checksum/crc8.h"
+#include "sps30.h"
+#include "xtimer.h"
+#include "byteorder.h"
+#include "kernel_defines.h"
+
+/**
+ * @name     SPS30 literal definitions
+ * @{
+ */
+#define SPS30_I2C_ADDR       (0x69) /**< Fixed I2C address for the Sensor */
+#define SPS30_CRC_POLY       (0x31) /**< Polynomial for the CRC calulation */
+#define SPS30_CRC_INIT       (0xFF) /**< Init value for the CRC calulation */
+#define SPS30_MESAURE_MODE   (0x03) /**< Fixed mode byte for start cmd */
+#define SPS30_DUMMY_BYTE     (0x00) /**< Fixed dummy byte for unused bytes */
+#define SPS30_PTR_LEN           (2) /**< Pointer address length in bytes */
+/** @} */
+
+/**
+* @name      Address pointer values for all SPS30 I2C commands
+* @{
+*/
+typedef enum {
+    SPS30_CMD_START_MEASURE   = 0x0010, /**< Start measurement mode */
+    SPS30_CMD_STOP_MEASURE    = 0x0104, /**< Stop measurement mode */
+    SPS30_CMD_RD_DATA_READY   = 0x0202, /**< Read data-ready flag */
+    SPS30_CMD_RD_MEASUREMENTS = 0x0300, /**< Read measured values */
+    SPS30_CMD_RW_AUTOCLEAN    = 0x8004, /**< Read/write autoclean interval */
+    SPS30_CMD_START_FAN_CLEAN = 0x5607, /**< Start fan cleaning */
+    SPS30_CMD_RD_ARTICLE      = 0xD025, /**< Read article code */
+    SPS30_CMD_RD_SERIAL       = 0xD033, /**< Read serial number */
+    SPS30_CMD_RESET           = 0xD304, /**< Reset */
+} sps30_cmd_t;
+/** @} */
+
+/**
+ * @brief    Combine payload and CRC into an SDS30 I2C data frame.
+ *
+ * @details  The CRC for `data` is calculated and written to `crcd_data`
+ *           together with the payload itself.
+ *           The format of `data` must be just the pure payload:
+ *           { data[0], data[1], data[2], ..., data[n]}
+ *           The format of `crcd_data` same as it shall be sent to the SPS30.
+ *           It consists of payload byte-pairs followed by a single CRC byte:
+ *           { data[0], data[1], csum[0],.., data[n-1], data[n], csum[n-1] }
+ *
+ * @param[in]    data        Source buffer containing just the raw payload
+ * @param[in]    len         Size of the 'data' buffer
+ * @param[out]   crcd_data   Destination buffer for combined payload and CRCs
+ *
+ * @pre          sizeof(crcd_data) must be equal to 1.5 * len
```

Comment:
may add an assert for that?

Reply:
not possible because there is no len parameter to the function. But the function is only used internally and the precondition should be true in all cases - that's why I opted for a precondition + doc instead of an explicit parameter.